xarray/core/groupby.py

```diff
@@ -882,9 +898,105 @@ def reduce_array(ar):
+        """ The dims of the resulting object (before any further reduction) """
+
+        for d in self.dims:
+            if d != self._group_dim:
+                yield d
+            # Grouping on a dimension
+            elif self._group_dim == self._group.name:
+                yield d
+            # Grouping on a coord that isn't a dimension
+            else:
+                yield self._group.name
+
+    def _npg_groupby(self, func: str):
+
+        # `.values` seems to be required for datetimes
+        indices, _ = pd.factorize(self._group.values)
+
+        array = npg_aggregate(self._obj, self._group_dim, func, indices)
+
+        # FIXME: Currently we're trying to use as much of the existing
+        # infrastructure as possible, but I'm struggling to fit it in — it may
+        # be easier to start from scratch.
+        #
+        # The existing model checks a single result `applied_example`; which in
+        # this case we don't have — we generate the whole array at once.
+
+        applied = applied_example = type(self._obj)(
+            data=array,
+            dims=tuple(self.dims_()),
+        )
+
+        # The remainder is mostly copied from `_combine`
+
+        # FIXME: this part seems broken at the moment — the `_infer_concat_args`
+        # doesn't return the correct result when the group isn't a dimensioned
+        # coordinate
+
+        coord = self._unique_coord
+        coord, dim, positions = self._infer_concat_args(applied_example)
+        # NB: These are commented out for simplicity.
+        # if shortcut:
+        #     combined = self._concat_shortcut(applied, dim, positions)
+        # else:
+        combined = concat(applied, dim)
+        combined = _maybe_reorder(combined, dim, positions)
+
+        if isinstance(combined, type(self._obj)):
+            # only restore dimension order for arrays
+            combined = self._restore_dim_order(combined)
+        # assign coord when the applied function does not return that coord
+        if coord is not None:  # and dim not in applied_example.dims:
+            # if shortcut:
+            #     coord_var = as_variable(coord)
+            #     combined._coords[coord.name] = coord_var
+            # else:
+            combined.coords[coord.name] = coord
+        combined = self._maybe_restore_empty_groups(combined)
+        combined = self._maybe_unstack(combined)
+
+        return combined
+
+    if OPTIONS["numpy_groupies"]:
+
+        def sum(self, dim=None):
+            grouped = self._npg_groupby(func="sum")
+            if dim:
+                return grouped.sum(dim)
+            else:
+                return grouped
+
+        def count(self, dim=None):
+            grouped = self._npg_groupby(func="count")
+            if dim:
+                return grouped.count(dim)
+            else:
+                return grouped
+
+        def mean(self, dim=None):
+            grouped = self._npg_groupby(func="mean")
+            if dim:
```

Comment:
Not right for mean. Would it work if we broadcast `indices` along all reduction dims before passing to npg?